messaging/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -109,15 +113,28 @@ public static Builder builder() {
         return new Builder();
     }
 
+    @SuppressWarnings("rawtypes")
     @Override
     public <R> Registration subscribe(String queryName,
                                       Type responseType,
                                       MessageHandler<? super QueryMessage<?, R>> handler) {
-        //noinspection rawtypes
-        CopyOnWriteArrayList<QuerySubscription> handlers =
-                subscriptions.computeIfAbsent(queryName, k -> new CopyOnWriteArrayList<>());
         QuerySubscription<R> querySubscription = new QuerySubscription<>(responseType, handler);
-        handlers.addIfAbsent(querySubscription);
+        List<QuerySubscription> handlers = subscriptions.computeIfAbsent(queryName, k -> new CopyOnWriteArrayList<>());
```

Comment:
I might've miss guided you a little in our offline discussion on this, so my apologies for that.

This change you've made seems to changes the behavior slightly.
Namely, not only are Query Handlers of the same name and response type in different Query Handling Components both subscribed (as already happened), but also handlers with the same name and response combination **within** a single Query Handling Component (QHCs for short).

The `handlers.addIfAbsent` only added a `QuerySubscription` if the `responseType` **and** `handler` are identical. The `responseType` is allowed to reoccur QHCs.
The `handler` actually **is** the QHC.

It is such a registration we should warn against. 
Differently put, where we can have a resolver for.


I think adjusting the `filter` operation to match the `handler` will do here.
Whether there's still a necessity to have a list of subscriptions for that use case...I'm not sure.

If we allow subscription of the same name-response combi between handlers instances, then we only need to match between the two handlers.
So, similar to how the `DuplicateCommandHandlerResolver` works.

Long story short, let me here your insights here, @Morlack.

Reply:
The `QHC` you are referring to here is the class. But the `handler` is a representation of the method (`MessageHandler`). If we would match on both `handler` and `responseType` it would always be unique, since methods are never registered twice. 

The difference between query handler resolution and command handler resolution is the relevance of return types.
For commands, only the command payload matters. Hence, a map of String -> CommandHandler is enough. For Query handlers, we also need to take the return type into account, hence filtering the list on that. 

I believe this is the same behavior. The `addIfAbsent` filtered on responseType+handler (which are unique), registering both. Later, in `SimpleQueryBus.query`, it loops over the handlers and gets the first valid result. This is why one of the handlers is lost silently. But they are both registered. 

Because of this I would argue that the behavior is the same